### PR TITLE
[FLINK-33663] Serialize CallExpressions into SQL

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR adds a `CallSyntax` as described in the FLIP and implementations for specific non standard builtin functions.

## Brief change log

* adds CallSyntax interface and changes in BuiltinFunctionDefinitions
* implements call syntax for functions in BuiltinFunctionDefinitions

## Verifying this change

Added tests in org.apache.flink.table.expressions.ExpressionSerializationTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/CallExpression.java [hunk 1]
```diff
@@ -216,6 +217,17 @@ public String asSummaryString() {
...
     }
 
+    @Override
+    public String asSerializableString() {
+        if (functionDefinition instanceof BuiltInFunctionDefinition) {
+            return ((BuiltInFunctionDefinition) functionDefinition)
+                    .getCallSyntax()
+                    .unparse(((BuiltInFunctionDefinition) functionDefinition).getSqlName(), args);
+        } else {
+            return CallSyntax.FUNCTION.unparse(getFunctionName(), args);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/CallExpression.java:227]: We should throw an exception when `functionIdentifier` is null. Something like "Only functions that have been registered before are serializable"?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/CallExpression.java [hunk 2]
```diff
@@ -216,6 +217,17 @@ public String asSummaryString() {
         return getFunctionName() + argList;
     }
 
+    @Override
+    public String asSerializableString() {
+        if (functionDefinition instanceof BuiltInFunctionDefinition) {
+            return ((BuiltInFunctionDefinition) functionDefinition)
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/CallExpression.java:223]: nit: use local variable and cast once

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/ValueLiteralExpression.java [hunk 3]
```diff
@@ -219,6 +219,11 @@ public String asSummaryString() {
         return stringifyValue(value);
     }
 
+    @Override
+    public String asSerializableString() {
+        return stringifyValue(value);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/ValueLiteralExpression.java:224]: are you sure that this is safe for all data types? A quick scan revealed that arrays could be an issue?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java [hunk 4]
```diff
@@ -69,23 +69,31 @@ public final class BuiltInFunctionDefinition implements SpecializedFunction {
 
     private final boolean isInternal;
 
+    private final CallSyntax callSyntax;
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:72]: Should we call this `SqlCallSyntax`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java [hunk 5]
```diff
@@ -327,12 +351,43 @@ public Builder internal() {
             return this;
         }
 
+        /**
+         * Overwrites the syntax used for unparsing a function into a SQL string. If not specified.
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:355]: nit: `If not specified,` also below

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 6]
```diff
@@ -406,6 +414,13 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
     public static final BuiltInFunctionDefinition IF =
             BuiltInFunctionDefinition.newBuilder()
                     .name("ifThenElse")
+                    .callSyntax(
+                            (sqlName, operands) ->
+                                    String.format(
+                                            "IF %s THEN %s ELSE %s",
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:420]: is that valid SQL?

Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:420]: Why not `asSerializableOperand`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 7]
```diff
@@ -679,6 +720,26 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
     public static final BuiltInFunctionDefinition DISTINCT =
             BuiltInFunctionDefinition.newBuilder()
                     .name("distinct")
+                    .callSyntax(
+                            (sqlName, operands) -> {
+                                final CallExpression callExpression =
+                                        (CallExpression) operands.get(0);
+                                if (callExpression.getFunctionDefinition()
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:727]: Can you add some comments here? This looks not straightforward compared to the other functions.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 8]
```diff
@@ -763,6 +827,7 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
     public static final BuiltInFunctionDefinition SUBSTR =
             BuiltInFunctionDefinition.newBuilder()
                     .name("substr")
+                    .callSyntax("SUBSTR", CallSyntax.SUBSTRING)
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:830]: Serialize into `SUBSTRING`? I don't understand why this synonym actually exists.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 9]
```diff
@@ -1904,6 +2055,25 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
...
+                            (sqlName, operands) -> {
+                                final Optional<String> fieldName =
+                                        ((ValueLiteralExpression) operands.get(1))
+                                                .getValueAs(String.class);
+
+                                return fieldName
+                                        .map(
+                                                n ->
+                                                        String.format(
+                                                                "%s.`%s`",
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2068]: not sure if the serialization format is correct here:
```
`%s`

```
 shouldn't this be properly escaped in the argument already

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 10]
```diff
@@ -2092,6 +2262,19 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
...
             BuiltInFunctionDefinition.newBuilder()
                     .name("IS_JSON")
+                    .callSyntax(
+                            (sqlName, operands) -> {
+                                final String s =
+                                        String.format(
+                                                "%s IS JSON",
+                                                CallSyntaxUtils.asSerializableOperand(
+                                                        operands.get(0)));
+                                if (operands.size() > 1) {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:2272]: can you add more comments for this special cases? It's hard to understand what is going on here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/CallSyntax.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+public interface CallSyntax {
+
+    String unparse(String sqlName, List<ResolvedExpression> operands);
+
+    default String unparseDistinct(String sqlName, List<ResolvedExpression> operands) {
+        throw new UnsupportedOperationException(
+                "Only the FUNCTION syntax supports the DISTINCT clause.");
+    }
+
+    CallSyntax FUNCTION =
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/CallSyntax.java:38]: add concrete examples to each syntax

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/CallSyntaxUtils.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+/** Utility functions that can be used for writing {@link CallSyntax}. */
+@Internal
+class CallSyntaxUtils {
+
+    /**
+     * Converts the given {@link ResolvedExpression} into a SQL string. Wraps the string with
+     * parenthesis if the expression is not a leaf expression such as e.g. {@link
+     * ValueLiteralExpression} or {@link FieldReferenceExpression}.
+     */
+    static String asSerializableOperand(ResolvedExpression expression) {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/CallSyntaxUtils.java:35]: why are not calling this method for all functions?

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/ExpressionSerializationTest.java [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+import java.util.stream.Stream;
+
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.table.api.Expressions.negative;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for serializing {@link BuiltInFunctionDefinitions} into a SQL string. */
+public class ExpressionSerializationTest {
+
+    public static Stream<TestSpec> testData() {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/ExpressionSerializationTest.java:57]: Will we also add a test in the planner that reuses the spec and checks whether the SQL is correct?
We could also extend the `BuiltInFunctionTestBase` and test serializability for all Table API expressions.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @dawidwys. I added some comments.
My biggest concern currently is SQL injection. We need to make sure that we properly escape where necessary.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 6 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:420]: what happens during nesting multiple ifs?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/ValueLiteralExpression.java (hunk 3 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/ValueLiteralExpression.java:224]: I am sure, it is not. My plan was to work on literal serialization next. Enabling e2e tests (in `BuiltinFunctionTestBase`) shows it actually makes sense to reverse the order and start with `ValueLiteralExpression` if we want to enable e2e tests along with this PR, I need to handle literals first.

FYI: not only arrays are the issue, but at least also temporals (TIMESTAMP, DATE, ...), and `null` value.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 8 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:830]: Neither do I. We have those two functions in Calcite as well. We have also two separate calls in Table API.

I'd rather remove one version from Table API if we want to serialize it into the same function.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 6 above)
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: > is that valid SQL?

It was wrong. Fixed it.

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/ExpressionSerializationTest.java (hunk 13 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/ExpressionSerializationTest.java:57]: I did the same for `BuiltInAggregateFunctionTestBase`

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request merged**